### PR TITLE
methods in custom ArrayList class implemented

### PR DESCRIPTION
**Pull request by TymurProkhorov**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,101 @@
...
+
+    private void resize() {
+        int newSize = (int) (array.length * INCREASE_SIZE);
+        Object[] resizedSize = new Object[newSize];
+        System.arraycopy(array, 0, resizedSize, 0, array.length);
+        this.array = resizedSize;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > this.size) {
```
Comment (apc10bbe) [inline src/main/java/core/basesyntax/ArrayList.java:97]: change  'if (index < 0 || index > this.size)'  to  'if (index < 0 || index >= size)'

**Review by apc10bbe — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,101 @@
...
+            resize();
+        }
+        array[this.size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index >= size) {
```
Comment (apc10bbe) [inline src/main/java/core/basesyntax/ArrayList.java:28]: here (ONLY ON THIS LINE) '=' not allowed, it will be  'if (index < 0 || index >= size)'

**Review by apc10bbe — CHANGES_REQUESTED** [review]

**Review by apc10bbe — APPROVED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,101 @@
...
+        this.size = 0;
+        this.array = new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (array.length == size) {
+            resize();
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/ArrayList.java:21]: You can move this conditional to method resize() and avoid code duplicates

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,101 @@
...
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (array.length == size) {
+            resize();
+        }
+        for (int i = size - 1; i >= index; i--) {
+            array[i + 1] = array[i];
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/ArrayList.java:36]: you can use System.arraycopy()

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,101 @@
...
+        System.arraycopy(array, index + 1, array, index, size - 1 - index);
+        size--;
+        return removedArray;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < array.length; i++) {
+            if (Objects.equals(array[i], element)) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/ArrayList.java:72]: read common mistakes. Don’t  use class Objects

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes 😉

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,96 @@
...
+        checkIndex(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (array.length == size) {
+            resize();
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/ArrayList.java:56]: Do you really need to resize the array when you remove the value?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Be attentive to fixing mistakes. Now you have a problem with tests.
image

**Review by liliia-ponomarenko — APPROVED** [review]
good job! 👍

**Pull request closed without merging**